### PR TITLE
prevent type mismatch from a go panic

### PR DESCRIPTION
**Pull request by amy**
https://github.com/rancher/rancher/issues/9483

### CODE REVIEW

**Comment by cjellick** [comment]
A reopened version of https://github.com/rancher/rancher-compose-executor/pull/72 (which got auto-closed when I delete the v1.6 branch)

File: config/schema_helpers.go [hunk 1]
```diff
@@ -60,27 +61,51 @@ func setupSchemaLoaders(schemaData string, schema *map[string]interface{}, schem
...
 	if val, ok := property["oneOf"]; ok {
-		validConditions := val.([]interface{})
-
-		for _, validCondition := range validConditions {
-			condition := validCondition.(map[string]interface{})
-			validTypes = append(validTypes, condition["type"].(string))
-		}
+		validTypes = oneOfValidTypes(val, validTypes)
 	} else if val, ok := property["$ref"]; ok {
+		logrus.Infof("REFERENCE %v\n", property["$ref"])
```
Comment (cjellick) [inline config/schema_helpers.go]: Let's bump this down to debug

Comment (amy) [reply config/schema_helpers.go]: I just removed it because it was a log that should have been left out of this PR.

**Comment by cjellick** [comment]
@amy can you add tests? I'm particularly concerned/interested in the logic in side this if block:
```
if _, ok := property["oneOf"]; !ok {
  ...
}
```

**Comment by amy** [comment]
The goal of these tests is to prove that invalid `docker-compose.yml` value types are handled smoothly without resulting in a Go panic. Previously, the Go panics should only have manifested in the `secrets` and `ulimits` fields.

Overview: Build the Rancher CLI. And provide types for fields that don't match those specified in docker-compose documentation. Flip the types for each field under `secrets` and `ulimits`

Secrets testing:
```
// expected output: You should receive an error from the CLI that tells you what the expected type is
    version: '2'
    services:
      test:
        image: nginx
        stdin_open: true
        tty: true
        secrets:
        - mode: '444' // this is an incorrect type for this field
          uid: '104'
          gid: '103'
          source: mylongsecret
          target: mylongsecretincontainer
    secrets:
      mylongsecret:
        external: true
```

```
// ulimits testing
version: '2'
services:
    test:
      image: nginx
      stdin_open: true
      tty: true
      ulimits:
        nproc: '65535' // this is an invalid type for this field
        nofile:
          soft: '20000' // this is an invalid type for this field
          hard: 40000
```

File: config/schema_helpers.go [hunk 2]
```diff
@@ -60,26 +60,49 @@ func setupSchemaLoaders(schemaData string, schema *map[string]interface{}, schem
 	return nil
 }
 
+func oneOfValidTypes(oneOf interface{}, validTypes []string) []string {
+	validConditions := oneOf.([]interface{})
```
Comment (cjellick) [inline config/schema_helpers.go]: i think you should add an `ok` check here to protect against panic from this assertion

File: config/schema_helpers.go [hunk 3]
```diff
@@ -60,26 +60,49 @@ func setupSchemaLoaders(schemaData string, schema *map[string]interface{}, schem
 	return nil
 }
 
+func oneOfValidTypes(oneOf interface{}, validTypes []string) []string {
```
Comment (cjellick) [inline config/schema_helpers.go]: This function name is not indicative of what it does. How about `appendValidType` or something like that? I want to communicate that it is appending the the slice passed in

File: config/schema_helpers.go [hunk 4]
```diff
@@ -91,5 +114,37 @@ func parseValidTypesFromSchema(schema map[string]interface{}, context string) []
...
+			property = properties[key].(map[string]interface{})
+		}
+
+		if _, ok := property["patternProperties"]; ok {
+			// address schema for 'ulimits'
+			patternProperties := make(map[string]interface{})
+			patternProperties = property["patternProperties"].(map[string]interface{})
+
+			regex := make(map[string]interface{})
+			if _, ok := patternProperties["^[a-z]+$"]; ok {
```
Comment (cjellick) [inline config/schema_helpers.go]: I don't think we should hardcode this pattern here. I think you can generalize it to something like (brace for bad psuedo-code):
```
for pattern, patternVal := range patternProperties {
  if val, ok patternVal["oneOf"]; ok {
    validTypes = oneOfValidTypes(val, validTypes)
  }
}
```

**Comment by amy** [comment]
Moved PR to: https://github.com/rancher/rancher-compose-executor/pull/86

**Pull request closed without merging**